### PR TITLE
Add diffing and animations to table views

### PR DESCRIPTION
**Pull request by Odrakir**
## What's this?
This adds a diffing algorithm so it's easy to animate changes to table view sources. It's a direct port of [RxDataSources](https://github.com/RxSwiftCommunity/RxDataSources). The diffing is not simple at all, but it includes many tests that guarantee that is working.

This PR also includes the necessary changes to make it work with our main log view. Sections have to be `Diffable` and Items have to be `Diffable` and `IEquatable` for this to work.

Finally I had to also edit the mapping of time entries which lead me to completely remove the class that took care of grouping them.

### Relationships
This is part of the grouping TE project - #4224. Doesn't close the issue though (the Android part still has to be confirmed).

## Why do we want this?
We want to have the animations back and also not having to worry about them.

## How is it done?
The collection from Foundation is still the same as it used to be (with the only different that the sections have to be diffable and the items also equatable).

Then in the iOS side the diffing is performed, where we find the singe changes every time there's a change in the collection. This get listed and passed to the table view for animation in a specific order that makes iOS happy.

### Why not another way?
It has to be done with an iOS specific diffing algorithm because of `UITableView` is picky. There are others, but this one seemed the easier one to port.

## Review hints
I wouldn't recommend spend too much time in the algorithm it self, do take a look at types, see if there's something that can be made more performant.

## :squid: Permissions
Not yet, this is still missing a performance test. And the ones about Flattening and Grouping have to be updated. And also, I want to update the docs on table view sources.

### CODE REVIEW

File: Toggl.Foundation.Tests/MvvmCross/Collections/Extensions/ChangesetExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                        if (destinationToSourceMapping.ContainsKey(itemPath))
+                        {
+                            var sourceIndex = destinationToSourceMapping[itemPath];
+                            resultItems.Add(original[sourceIndex.sectionIndex].Items[sourceIndex.itemIndex]);
+                        }
+                        else
+                        {
+                            if (!findNextUntouchedSourceItem())
+                            {
+                                //TODO THROW fatalError("Oooops, wrong commands.")
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/Collections/Extensions/ChangesetExtensions.cs]: 👀

File: Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 namespace Toggl.Foundation.Tests.MvvmCross.Transformations
 {
+    /*
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs]: Disabling all of those tests? 👀 The tests should be merged the same way the transformation classes were merged.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+                            Console.WriteLine("Throw");
+                            break;
+                    }
+                }
+
+                var newSection = new TSection();
+                newSection.Initialize(initialSections[i].Header, afterDeleteItems);
+                afterDeleteState.Add(newSection);
+            }
+            // }
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👀

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 4]
```diff
@@ -0,0 +1,767 @@
...
+        }
+
+        private List<Changeset> generateDeleteSectionsDeletedItemsAndUpdatedItems()
+        {
+            var deletedSections = new List<int>();
+            var deletedItems = new List<ItemPath>();
+            var updatedItems = new List<ItemPath>();
+            var afterDeleteState = new List<TSection>();
+
+            // mark deleted items {
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: is this supposed to be a `#region Mark deleted items`?

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+            MoveIndex = moveIndex;
+        }
+
+        public static ItemAssociatedData Initial()
+        {
+            return new ItemAssociatedData(EditEvent.Untouched, null, null);
+        }
+    }
+
+    public class Diffing<TSection, THeader, TElement>
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: This file is huge. I think it would make sense to move all those classes and enums and whatnot into separate files - it would make this file 100 lines shorter.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+    {
+        public long DuplicatedIdentity { get; }
+
+        public DuplicateSectionException(long identity)
+        {
+            DuplicatedIdentity = identity;
+        }
+    }
+
+    public class ItemPath : IEquatable<ItemPath>
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `internal`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 7]
```diff
@@ -0,0 +1,767 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Toggl.Foundation.MvvmCross.Collections;
+
+namespace Toggl.Daneel.ViewSources.Generic
+{
+    public class DuplicateItemException : Exception
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `internal`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+    {
+        public long DuplicatedIdentity { get; }
+
+        public DuplicateItemException(long identity)
+        {
+            DuplicatedIdentity = identity;
+        }
+    }
+
+    public class DuplicateSectionException : Exception
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `internal`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                return (sectionIndex * 397) ^ itemIndex;
+            }
+        }
+    }
+
+    enum EditEvent
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `internal`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+        Inserted,
+        InsertedAutomatically,
+        Deleted,
+        DeletedAutomatically,
+        Moved,
+        MovedAutomatically,
+        Untouched
+    }
+
+    class SectionAssociatedData
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `internal`

File: Toggl.Daneel/Extensions/Reactive/UITableViewExtensions.cs [hunk 11]
```diff
@@ -10,24 +16,115 @@ namespace Toggl.Daneel.Extensions.Reactive
...
+                {
+                    dataSource.SetSections(finalSections);
+                    reactive.Base.ReloadData();
+                    return;
+                }
+
+                var diff = new Diffing<TSection, THeader, TModel>(initialSections, finalSections);
+                var changeset = diff.computeDifferences();
+
+                foreach (var difference in changeset)
```
Comment (simonrozsival) [inline Toggl.Daneel/Extensions/Reactive/UITableViewExtensions.cs]: Please make it obvious to the reader of the code why the changes are applied in a loop and not all at once. A comment would be enough.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 12]
```diff
@@ -0,0 +1,767 @@
...
+                        var initialData = itemAssociatedData[j];
+
+                        if (initialData.EditEvent == EditEvent.Deleted)
+                        {
+                            continue;
+                        }
+
+                        if (initialData.MoveIndex == null)
+                        {
+                            // TODO throw try precondition(false, "Item was moved, but no final location.")
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👀

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 13]
```diff
@@ -0,0 +1,767 @@
...
+                    }
+
+                    var newSection = new TSection();
+                    newSection.Initialize(section.Header, items);
+                    var modifiedSection = newSection;
+
+                    return modifiedSection;
+                }
+                else {
+                    // TODO Throw try precondition(false, "This is weird, this shouldn't happen")
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👀

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 14]
```diff
@@ -0,0 +1,767 @@
...
+        {
+            Dictionary<long, int> indexedSections = new Dictionary<long, int>();
+
+            for (int i = 0; i < sections.Count(); i++)
+            {
+                var section = sections[i];
+
+                if (indexedSections.ContainsKey(section.Identity))
+                {
+                    // TODO EXCEPTION  print("Section \(section) has already been indexed at \(indexedSections[section.identity]!)")
```
Comment (simonrozsival) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👀

File: Toggl.Foundation.Tests/MvvmCross/Collections/Extensions/ChangesetExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+        private static List<TSection> applySectionMovesAndInserts<TSection, THeader, TElement>(
+            this Diffing<TSection, THeader, TElement>.Changeset changeset,
+            List<TSection> original
+        )
+            where TSection : IAnimatableSectionModel<THeader, TElement>, new()
+            where THeader : IDiffable
+            where TElement : IDiffable, IEquatable<TElement>
+        {
+            if (changeset.UpdatedSections.Count != 0) {
+                // TODO Throw fatalError("Section updates aren't supported")
```
Comment (simonrozsival) [inline Toggl.Foundation.Tests/MvvmCross/Collections/Extensions/ChangesetExtensions.cs]: 👀

File: Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs (hunk 2 above)
Comment (Odrakir) [reply Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs]: I tried to merge some of them with the flattening tests, but other don't make sense anymore. Please, do take a look if you find something is missing.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs (hunk 4 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: It can be 🔥 , just copied like it was in the Swift code.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs (hunk 5 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👌

File: Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs (hunk 2 above)
Comment (simonrozsival) [reply Toggl.Foundation.Tests/MvvmCross/Transformations/TimeEntriesGroupingTests.cs]: ok

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs (hunk 7 above)
Comment (kornelijepetak) [reply Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `sealed`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs (hunk 8 above)
Comment (kornelijepetak) [reply Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `sealed`

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 16]
```diff
@@ -0,0 +1,767 @@
...
+            return result;
+        }
+
+        private static (List<SectionAssociatedData>, List<SectionAssociatedData>) calculateSectionMovements(
+            List<TSection> initialSections, List<TSection> finalSections)
+        {
+            var initialSectionIndexes = indexSections(initialSections);
+
+            var initialSectionData = Enumerable.Range(0, initialSections.Count)
+                .Select(_ => SectionAssociatedData.Initial()).ToList();
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: Please move `ToList()` on the new line. It's an enumerable operator 😉

Payback time 😛

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 17]
```diff
@@ -0,0 +1,767 @@
...
+            var finalSectionData = Enumerable.Range(0, finalSections.Count)
+                .Select(_ => SectionAssociatedData.Initial()).ToList();
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]:
```suggestion
                .Select(_ => SectionAssociatedData.Initial())
                .ToList();
```

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 18]
```diff
@@ -0,0 +1,767 @@
...
+                initialSectionData[i].IndexAfterDelete = sectionIndexAfterDelete;
+                sectionIndexAfterDelete += 1;
+            }
+
+            // moved sections
+
+            int? untouchedOldIndex = 0;
+            int? findNextUntouchedOldIndex(int? initialSearchIndex)
+            {
+                if (!initialSearchIndex.HasValue) {
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `Brace on the new line` policy violation.

But otherwise, since it's as simpl as `return null`, I'd ditch the braces completely in this simple case.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 19]
```diff
@@ -0,0 +1,767 @@
...
+            }
+
+            // inserted and moved sections {
+            // this should fix all sections and move them into correct places
+            // 2nd stage
+            for (var i = 0; i < finalSections.Count; i++)
+            {
+                untouchedOldIndex = findNextUntouchedOldIndex(untouchedOldIndex);
+
+                // oh, it did exist
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: 👀

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 20]
```diff
@@ -0,0 +1,767 @@
...
+
+                initialSectionData[initialSectionIndex].MoveIndex = i;
+                finalSectionData[i].MoveIndex = initialSectionIndex;
+            }
+
+            var sectionIndexAfterDelete = 0;
+
+            // deleted sections
+
+            for (var i = 0; i < initialSectionData.Count(); i++)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: It's a list, use the Count property.

```suggestion
            for (var i = 0; i < initialSectionData.Count; i++)
```

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 21]
```diff
@@ -0,0 +1,767 @@
...
+                    i++;
+                }
+
+                return null;
+            }
+
+            // inserted and moved sections {
+            // this should fix all sections and move them into correct places
+            // 2nd stage
+            for (var i = 0; i < finalSections.Count; i++)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: Any specific reason why is this not a `foreach`?

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 22]
```diff
@@ -0,0 +1,767 @@
...
+                    initialSectionData[oldSectionIndex.Value].EditEvent = moveType;
+                }
+                else
+                {
+                    finalSectionData[i].EditEvent = EditEvent.Inserted;
+                }
+            }
+
+            // inserted sections
+            for (var i = 0; i < finalSectionData.Count; i++)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: This whole `for` loop can be simplified to a more functional form:
```cs
finalSectionData
    .Where(section => section.MoveIndex.HasValue)
    .ForEach(section => section.EditEvent = EditEvent.Inserted);
```

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 23]
```diff
@@ -0,0 +1,767 @@
...
+
+        private IEnumerable<Changeset> generateInsertAndMoveSections()
+        {
+            var movedSections = new List<(int, int)>();
+            var insertedSections = new List<int>();
+
+            for (int i = 0; i < initialSections.Count; i++)
+            {
+
+                switch (initialSectionData[i].EditEvent) {
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: Lot's of egyptian brackets in this file.

File: Toggl.Foundation.MvvmCross/Collections/Diffing.cs [hunk 24]
```diff
@@ -0,0 +1,767 @@
...
+                    //TODO THROW try precondition(false, "Unhandled case in initial sections")
+                    Console.WriteLine("Throw");
+                    break;
+                }
+            }
+
+            for (int i = 0; i < finalSections.Count; i++)
+            {
+                switch (finalSectionData[i].EditEvent) {
+                case EditEvent.Inserted:
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/Diffing.cs]: `case`s are by default indented one step.
Also, the brace should go on a newline.

File: Toggl.Daneel/Extensions/Reactive/UITableViewExtensions.cs (hunk 11 above)
Comment (kornelijepetak) [reply Toggl.Daneel/Extensions/Reactive/UITableViewExtensions.cs]: Yes, especially to us non-iOS devs, it's a mystery.
